### ISSUE TITLE
Multiple Signature Help is not provided

### ISSUE THREAD

**Issue by nadeeshaan** [ballerina-platform/ballerina-lang#35233]
**Description:**
Consider the following source snippet,

```
class Cl1 {
    function init(int arg1, string arg2) {
        
    }
}

class Cl2 {
    function init(string arg1, string arg2) {
        
    }
}

public function main() {
    Cl1 | Cl2 cl = new(<cursor>);
}

```

- At the cursor position, we should get two signature help options for each of the classes.

**Affected Versions:**
2201.0.1-SNAPSHOT at least

**Comment by DamithaSenevirathne** [comment]
During the type check process current compiler can not identify the correct type in similar cases. Because of this reason we can not infer the correct type and it returns following  error `cannot infer type of the object from '(Cl1|Cl2)'`

**Comment by DamithaSenevirathne** [comment]
Blocked by #35600

**Comment by IMS94** [comment]
@DamithaSenevirathne can't we use the `ContextTypeResolver` here to derive the expected type of the `new()` expression?

**Comment by DamithaSenevirathne** [comment]
@IMS94 in the current implementation of `ContextTypeResolver` nodeAtCursor is identified as `ImplicitNewExpressionNode` and  inside the transform method we are extracting classSymbol as follows

```java
Optional<TypeSymbol> classSymbol = context.currentSemanticModel()
                .flatMap(semanticModel -> semanticModel.typeOf(implicitNewExpressionNode))
                .flatMap(typeSymbol -> Optional.of(CommonUtil.getRawType(typeSymbol))).stream().findFirst();
```

But here ` context.currentSemanticModel().flatMap(semanticModel -> semanticModel.typeOf(implicitNewExpressionNode))`

returns `BallerinaCompiationErrorTypeSymbol`, So we can not use the current `ContextTypeResolver` to derive the expected type

**Comment by nadeeshaan** [comment]
If we are to support this behaviour, we need to introduce a fail-safe `ContextTypeResolver` which can deal with the compilation errors by falling back to a known set of error scenarios and using alternative type search mechanisms.

**Comment by IMS94** [comment]
@nadeeshaan yes. Right now @DamithaSenevirathne is working on an expected type API from semantic API side (#36539) since semantic API has access to the expected type information on BLang nodes + expected type is required for the debugger and ls extensions. We can decide on an approach once that API is provided.

**Comment by IMS94** [comment]
Since the expected type API is now merged, we can revisit this.
